lib/Serialization/Deserialization.cpp

```diff
@@ -5391,7 +5401,7 @@ DeclDeserializer::readAvailable_DECL_ATTR(SmallVectorImpl<uint64_t> &scratch,
     platformAgnostic = PlatformAgnosticAvailabilityKind::Deprecated;
   else if (isNoAsync)
     platformAgnostic = PlatformAgnosticAvailabilityKind::NoAsync;
-  else if (((PlatformKind)platform) == PlatformKind::none &&
+  else if (platform == PlatformKind::none &&
```

Comment:
The original code broke a general rule of Swift serialization. We should always use a locked down version of enums for things being serialized to avoid this kind of side effect from other compiler changes. See the `AccessLevel` enum in `ModuleFormat.h` for an example.

This new check is an improvement valuable on its own, and considering this specific use case avoiding the required duplication makes sense. However we might still get into an issue where the reader compiler interprets the platform differently from the writer. The check on the vendor field may be enough to catch it in most cases.